sway/desktop/xwayland.c

```diff
@@ -197,6 +215,35 @@ static void set_fullscreen(struct sway_view *view, bool fullscreen) {
 	wlr_xwayland_surface_set_fullscreen(surface, fullscreen);
 }
 
+static bool wants_floating(struct sway_view *view) {
+	// TODO:
+	// We want to return true if the window type contains any of these:
+	// NET_WM_WINDOW_TYPE_DIALOG
+	// NET_WM_WINDOW_TYPE_UTILITY
+	// NET_WM_WINDOW_TYPE_TOOLBAR
+	// NET_WM_WINDOW_TYPE_SPLASH
+	//
+	// We also want to return true if the NET_WM_STATE is MODAL.
+	// wlroots doesn't appear to provide all this information at the moment.
+	struct wlr_xwayland_surface *xsurface = view->wlr_xwayland_surface;
+	uint32_t *atom = xsurface->window_type;
+	for (size_t i = 0; i < xsurface->window_type_len; ++i) {
+		wlr_log(L_DEBUG, "xwayland window type %i", *atom);
+		// TODO: Come up with a better way of doing this
+		switch (*atom) {
+		case 36: // NET_WM_WINDOW_TYPE_UTILITY
+		case 44: // NET_WM_WINDOW_TYPE_SPLASH
+		case 276: // ? PGP passphrase dialog
+		case 337: // ? Firefox open file dialog
+		case 338: // ? Firefox open file dialog
```

Comment:
Let's implement this instead of having these magic values :)

Reply:
wlroots needs to expose its atom names first. And I'm a bit confused with how this all works - wlroots lacks atom names `DIALOG` and `TOOLBAR` which are used by i3. And according to `wlr_xwayland_surface_is_unmanaged`, surfaces with atoms `SPLASH` and `UTILITY` will be unmanaged, but to match i3 these would need to be managed and floated.

BTW, according to Ongy the atom values are defined at runtime, so I've removed the hard coded values from my PR. Keeping them there may cause false positives.

Reply:
I can do the wlroots PR tonight if that helps